File: 20210523LangReview/Python/08-interface/interface.js
/*

this module is to demonstrate interface concept in JavaScript

@ZL, 20210526

[ conclusion ]
===
- JavaScript does not support interface
- but it does have a walkaround, using class

*/

class Pack {
    start() {}
}

class Carton extends Pack {
    constructor(name, age) {
        super();
        this.name = name;
        this.age =  age;
    }

    start() {
        return "class Carton says hello";
    }
}

let c = new Carton("TS", 30);
print(c.start());

File: 20210523LangReview/Python/08-interface/interface.py
"""

this module is to demonstrate interface concept in Python

@ZL, 20210526

[ conclusion ]
===
- Python does not support interface
- but it does have a walkaround, using class
===

"""

class Pack:
    def start(self):
        pass

class Carton(Pack):
    def __init__(self, *args):
        self.args = args
    
    def __repr__(self):
        return "Carton{!r}".format(self.args)

    def start(self):
        return "class Carton says hello"

if __name__ == "__main__":
    c: Carton = Carton("TS", 30)
    print(c)
    print(c.start())